File: leetcode/easy/171_excel_sheet_column_number/dduck_171_excel_sheet_column_number.py
# https://leetcode.com/problems/excel-sheet-column-number

from unittest import TestCase

# Runtime: 32 ms, faster than 53.72% of Python3 online submissions for Excel Sheet Column Number.
# Memory Usage: 12.8 MB, less than 100.00% of Python3 online submissions for Excel Sheet Column Number.
# Runtime: 28 ms, faster than 81.16% of Python3 online submissions for Excel Sheet Column Number.
# Memory Usage: 12.7 MB, less than 100.00% of Python3 online submissions for Excel Sheet Column Number.

class Solution:
    def titleToNumber(self, s: str) -> int:
        sum = 0
        for i, v in enumerate(s[::-1]):
            column_number = ord(v) % 65 + 1
            exponent = (26 ** i)
            sum += exponent * column_number or column_number
        return sum


TestCase().assertEqual(Solution().titleToNumber('A'), 1)
# Input: "A"
# Output: 1

TestCase().assertEqual(Solution().titleToNumber('AA'), 27)
TestCase().assertEqual(Solution().titleToNumber('AAA'), 703)
#
TestCase().assertEqual(Solution().titleToNumber('AB'), 28)
# # Input: "AB"
# # Output: 28
#
TestCase().assertEqual(Solution().titleToNumber('ZY'), 701)
# Input: "ZY"
# Output: 701

File: leetcode/easy/171_excel_sheet_column_number/dduck_171_excel_sheet_column_number.js
// https://leetcode.com/problems/excel-sheet-column-number

// Runtime: 60 ms, faster than 98.77% of JavaScript online submissions for Excel Sheet Column Number.
//   Memory Usage: 35 MB, less than 73.33% of JavaScript online submissions for Excel Sheet Column Number.
// Runtime: 68 ms, faster than 84.71% of JavaScript online submissions for Excel Sheet Column Number.
//   Memory Usage: 34.8 MB, less than 100.00% of JavaScript online submissions for Excel Sheet Column Number.
// Runtime: 76 ms, faster than 55.89% of JavaScript online submissions for Excel Sheet Column Number.
//   Memory Usage: 34.9 MB, less than 80.00% of JavaScript online submissions for Excel Sheet Column Number.

const titleToNumber = (s) => s.split('')
  .reduceRight((a, c, i, arr) => {
    const exponent = arr.length - i - 1;
    const columnNumber = (c.charCodeAt(0) % 65) + 1;
    return a + ((26 ** exponent) * columnNumber || columnNumber);
  }, 0);

// Runtime: 72 ms, faster than 70.47% of JavaScript online submissions for Excel Sheet Column Number.
//   Memory Usage: 35.1 MB, less than 73.33% of JavaScript online submissions for Excel Sheet Column Number.
// Runtime: 80 ms, faster than 40.25% of JavaScript online submissions for Excel Sheet Column Number.
//   Memory Usage: 35 MB, less than 73.33% of JavaScript online submissions for Excel Sheet Column Number.
const titleToNumber2 = (s) => s.split('')
  .reverse()
  .map((v, i) => {
    const columnNumber = (v.charCodeAt(0) % 65) + 1;
    return (26 ** i) * columnNumber || columnNumber;
  })
  .reduce((a, c) => a + c);

console.log(titleToNumber('A'), 1);
// Input: "A"
// Output: 1

console.log(titleToNumber('AA'), 27);
console.log(titleToNumber('AAA'), 703);

console.log(titleToNumber('AB'), 28);
// Input: "AB"
// Output: 28

console.log(titleToNumber('ZY'), 701);
// Input: "ZY"
// Output: 701
